### PR TITLE
Load SafeTensor weights and handle DTensor sharded loading to enable distributed models to run with real weights

### PR DESCRIPTION
**Pull request by lessw2020**
This PR:
1 - adds safetensor_utils.py with supporting functions to enable per stage loading of HF Safetensor weights into the new distributed models.  All weights for the example 7B-HF-chat are now loading into their respective stages. 

2 - Because distributed now has built in parallelism, adds a dtensor_utils to appropriately shard and load weights into DTensor portions of the model. 

~~~
[rank2]:08-26 22:17:18.477 - distributed.safetensor_utils:281 - INFO - Ignoring 34 missing cache, freqs, mask layers
[rank2]:08-26 22:17:18.477 - distributed.safetensor_utils:293 - INFO - Fully updated state dict.
[rank2]:08-26 22:17:18.477 - distributed.safetensor_utils:294 - INFO - Loaded 146 weights into stage module
[rank2]:08-26 22:17:18.481 - __main__:40 - INFO - Success - Loaded 146 weights, 0 missing weights
~~~
and
~~~
[rank0]:08-26 22:17:18.448 - distributed.safetensor_utils:281 - INFO - Ignoring 34 missing cache, freqs, mask layers
[rank0]:08-26 22:17:18.448 - distributed.safetensor_utils:293 - INFO - Fully updated state dict.
[rank0]:08-26 22:17:18.448 - distributed.safetensor_utils:294 - INFO - Loaded 145 weights into stage module
[rank0]:08-26 22:17:18.452 - __main__:40 - INFO - Success - Loaded 145 weights, 0 missing weights
~~~

3 - modularizes dist_run to compartmentalize the respective functions for creating device mesh, loading weights, etc.

4 - adds a new logging utils setup (sample shown above) to enable us to see rank/time/file/line number info for better debugging. 

5 - adds a modified run_dist.sh which enables specifying gpus to run on (for shared servers), port number, etc.

6 - adds a variety of checks to ensure that final model has all expected dtype, all stages are on device and not cpu, etc. 

7 - adds a dtype tracker to weights lookup dict so that we can autohandle PipelineStage creation. (resolves https://github.com/pytorch/torchchat/issues/1087)

Screenshot 2024-08-29 at 11 55 02 AM

### CODE REVIEW

**Comment by lessw2020** [comment]
the failing job is unrelated to this PR - it appears it's timing out on a gguf weight save:
~~~
HTTP request sent, awaiting response... 200 OK
Length: 3825807040 (3.6G) [binary/octet-stream]
Saving to: ‘gguf_files/llama-2-7b.Q4_0.gguf’
~~~

File: build/model_dist.py [hunk 1]
```diff
@@ -108,6 +111,7 @@ def forward(self, x: Tensor, input_pos: Optional[Tensor] = None) -> Tensor:
 
         if self.stage_idx == self.n_stages - 1:
             x = self.norm(x)
+            logger.info(f"final stage output info: {x.dtype=}, {x.shape=}, {x.device=}\n {self.output.weight.dtype=}, {self.output.weight.shape=},")
```
Comment (kwen2501) [inline build/model_dist.py:114]: maybe comment out the logging? 
torch compile or export may graph break here if without special flags.

File: build/model_dist.py [hunk 2]
```diff
@@ -290,5 +294,7 @@ def _norm(self, x):
         return x * torch.rsqrt(torch.mean(x * x, dim=-1, keepdim=True) + self.eps)
 
     def forward(self, x: Tensor) -> Tensor:
+        if x.device != self.weight.device:
+            logger.info(f"RMSNorm weight mismatch -  input shape: {x.shape}, {x.device=}, self.weight.device={self.weight.device}")
```
Comment (kwen2501) [inline build/model_dist.py:298]: same

File: dist_run.py [hunk 3]
```diff
@@ -6,67 +6,151 @@
...
-
-from build.model import ModelArgs
+from build.model import TransformerArgs
 from build.model_dist import TransformerStage
+from distributed.logging_utils import setup_logging
+from distributed.safetensor_utils import (
+    get_hf_config_file,
+    get_hf_weight_map_and_path,
+    load_safetensor_weights,
+)
```
Comment (kwen2501) [inline dist_run.py:19]: nit: if `get_hf_config_file` etc are not related to distributed, maybe `safetensor_utils` can be moved to a separate sub-package? Can be in a different PR though.

File: dist_run.py [hunk 4]
```diff
@@ -6,67 +6,151 @@
...
+from distributed.dtensor_utils import find_cpu_tensors, record_module_dtypes
+from distributed.utils import Color as color
```
Comment (kwen2501) [inline dist_run.py:21]: nit: move `Color` out of `distributed.utils`?

File: dist_run.py [hunk 5]
```diff
@@ -6,67 +6,151 @@
...
+def _init_distributed():
+    dist.init_process_group("nccl")
+    rank = dist.get_rank()
+    world_size = dist.get_world_size()
+    torch.cuda.set_device(rank)
+    return rank, world_size
+
+
+def _create_device_mesh(mesh_dimensions):
+    return dist.init_device_mesh("cuda", mesh_dimensions, mesh_dim_names=("pp", "tp"))
```
Comment (kwen2501) [inline dist_run.py:38]: nit: `_create_device_mesh` is not necessarily more readable than `init_device_mesh` while they both cost 1 line :)

File: dist_run.py [hunk 6]
```diff
@@ -6,67 +6,151 @@
...
+    )
+    logger.info(
+        f"Success - Loaded {num_loaded_weights} weights, {num_missing_weights} missing weights"
+    )
+    if num_missing_weights > 0:
+        raise ValueError(f"Missing {num_missing_weights} weights")
+
+
+def _cleanup():
+    dist.barrier()
```
Comment (kwen2501) [inline dist_run.py:54]: nit: how about commenting out the barrier, and add a comment saying that "Uncomment it if you hit a hang"?

File: dist_run.py [hunk 7]
```diff
@@ -6,67 +6,151 @@
...
-    rank = dist.get_rank()
     device = torch.device(f"cuda:{rank}")
 
-    # Create parallel model with device_mesh context
     with device:
         with tp_mesh:
             model = TransformerStage(config, pp_rank, nstages)
             model.setup_caches(1, 4096)
+            # TODO: refine this .to once we start using fp8 for KV cache
+            model = model.to(model_dtype)
```
Comment (kwen2501) [inline dist_run.py:87]: nit: 
1. see if `config` provides an option for dtype.
2. Note that `build.utils` has a function named `get_precision()`. It seems to me that's where all precisions are controlled.
Screenshot 2024-08-29 at 3 39 43 PM

File: dist_run.py [hunk 8]
```diff
@@ -6,67 +6,151 @@
...
 
-    # Run pipeline
-    schedule = ScheduleGPipe(stage, mbs)
+    cpu_tensors = find_cpu_tensors(stage.submod)
+    if len(cpu_tensors) > 0:
+        raise ValueError("Found cpu tensors in stage")
+
+    # Load weights
+    logger.info(f"Loading weights for {pp_rank=} on {device=}")
+    _load_model_weights(stage.submod, hf_model_name, device=device, logger=logger)
```
Comment (kwen2501) [inline dist_run.py:116]: Shall we _load_model_weights before creating pipeline stage?

File: dist_run.py [hunk 9]
```diff
@@ -6,67 +6,151 @@
...
+
+    stage.submod.eval()
```
Comment (kwen2501) [inline dist_run.py:118]: Same.

File: dist_run.py [hunk 10]
```diff
@@ -6,67 +6,151 @@
...
+
+    # this check confirms that there are no cpu tensors in the model..we expect this to be true.
+    cpu_tensors = find_cpu_tensors(stage.submod)
+    # logger.info(f"Found {len(cpu_tensors)} cpu tensors: {cpu_tensors}")
+    if len(cpu_tensors) > 0:
+        raise ValueError("Found cpu tensors in stage")
```
Comment (kwen2501) [inline dist_run.py:124]: duplicate code?

File: dist_run.py [hunk 11]
```diff
@@ -6,67 +6,151 @@
...
+
+    # verify dtypes for model - expect all to be model_dtype except for bool causal_mask atm.
+    dtype_count, dtype_locations, fp32_locations = record_module_dtypes(stage.submod)
+    logger.info(
+        f"Stage Dtypes - Found {len(dtype_count)} dtypes: {dtype_count.items()}"
+    )
+    assert (
+        len(dtype_count) == 2
+    ), f"Expected 2 dtypes in model after checkpoint loading: {model_dtype} and {torch.bool}"
```
Comment (kwen2501) [inline dist_run.py:133]: nit: comment out this block?

File: distributed/dtensor_utils.py [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+            num_shards = mesh.size(0) # Assuming sharding is always along the first mesh dimension
+            shard_size = weight_tensor.size(shard_dim) // num_shards
+            shard_index = mesh.get_coordinate()[0]
+            
+            start_idx = shard_index * shard_size
+            end_idx = start_idx + shard_size
+            
+            slice_list = [slice(None)] * weight_tensor.dim()
+            slice_list[shard_dim] = slice(start_idx, end_idx)
+            weight_tensor = weight_tensor[tuple(slice_list)]
```
Comment (kwen2501) [inline distributed/dtensor_utils.py:105]: nit: can you add some comments for this block please? Thanks!

File: distributed/dtensor_utils.py [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+        local_tensor = new_dtensor.to_local()
+        local_shard_shape = local_tensor.shape
+        global_shape = new_dtensor.shape
+        logger.info("=" * 50)
+        logger.info(f"New DTensor: {global_shape=}, Local shard shape: {local_shard_shape}, Placements: {placements}")
+        model_shard_shape = model_dtensor.to_local().shape
+        model_global_shape = model_dtensor.shape
+        logger.info(f"Model DTensor: {model_global_shape=}, Local shard shape: {model_shard_shape}, Placements {model_dtensor.placements}")
+        assert local_shard_shape == model_shard_shape, f"Local shard shape {local_shard_shape} does not match model shard shape {model_shard_shape}"
+        logger.info("=" * 50)
```
Comment (kwen2501) [inline distributed/dtensor_utils.py:129]: nit: simplify this debug part a bit?

File: distributed/dtensor_utils.py [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+    
+    # model_dtensor.copy_(new_dtensor)
+    return new_dtensor
+
+
+def inspect_dtensor_sharding(dtensor):
+    """ hepful debug util for inspecting DTensor sharding """
+    if not is_dtensor(dtensor):
+        logger.info(f"This tensor {dtensor} is not a DTensor")
```
Comment (kwen2501) [inline distributed/dtensor_utils.py:138]: nit: logger.debug? Same below.

File: distributed/force_download.py
Comment (kwen2501) [inline distributed/force_download.py:1]: Where is this file used?

File: distributed/safetensor_utils.py [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+    file: str,
+    updated_states: Set[str],
+    device: torch.device,
+):  
+    count_dtensors_loaded = 0
+    for param, file_with_param in weight_map.items():
+        if file_with_param == file and param in state_dict:
+            model_param = (
+                "output.weight" if param == "output.weight" else f"model.{param}"
+            )
```
Comment (kwen2501) [inline distributed/safetensor_utils.py:248]: nit: is this part specific ?

File: distributed/safetensor_utils.py [hunk 16]
```diff
@@ -0,0 +1,325 @@
...
+                
+                state_dict[param] = model_tensor
+                count_dtensors_loaded += 1
+                
+            else:
+                # regular tensor, just update directly
+                checkpoint_tensor = checkpoint_tensor.to(device)
+                state_dict[param] = checkpoint_tensor
+
+            state_dict[param] = state_dict[param].to(checkpoint_tensor.dtype)
```
Comment (kwen2501) [inline distributed/safetensor_utils.py:277]: wondering the purpose of this to() call. 
It seems we just updated `state_dict` with `checkpoint_tensor` above.

File: distributed/safetensor_utils.py [hunk 17]
```diff
@@ -0,0 +1,325 @@
...
+            
+            assert state_dict[param].dtype == checkpoint_tensor.dtype
+            assert state_dict[param].dtype == torch.float16, f"{param} dtype is not fp16"
```
Comment (kwen2501) [inline distributed/safetensor_utils.py:280]: is this line specific?

File: distributed/safetensor_utils.py [hunk 18]
```diff
@@ -0,0 +1,325 @@
...
+
+def clean_cache_keys(input_set: Set[str]) -> Set[str]:
+    """Remove cache, freqs, mask params from checkpoint update set...we expect these to be generated """
+    return {
+        item
+        for item in input_set
+        if not (item.endswith("cache") or item in ["freqs_cis", "causal_mask"])
+    }
+
+def handle_missing_keys(
```
Comment (kwen2501) [inline distributed/safetensor_utils.py:300]: nit: comment for this function?

File: distributed/safetensor_utils.py [hunk 19]
```diff
@@ -0,0 +1,325 @@
...
+        missing_keys = clean_cache_keys(missing_keys)
+        after_len = len(missing_keys)
+        if after_len < start_len:
+            logger.info(
+                f"Ignoring {start_len - after_len} missing cache, freqs, mask layers"
+            )
+    return missing_keys
+
+
+def log_loading_status(missing_keys: Set[str], updated_states: Set[str]):
```
Comment (kwen2501) [inline distributed/safetensor_utils.py:318]: nit: this function is used at only one place, we can inline it.

**Review by kwen2501 — APPROVED** [review]
Thanks for getting DTensor loading to work!

File: build/model_dist.py [hunk 20]
```diff
@@ -91,13 +93,14 @@ def setup_caches(self, max_batch_size, max_seq_length):
...
         self.register_buffer("causal_mask", causal_mask, persistent=True)
-
+            
     def forward(self, x: Tensor, input_pos: Optional[Tensor] = None) -> Tensor:
         assert self.freqs_cis is not None, "Caches must be initialized first"
         if input_pos is None:
             input_pos = torch.arange(x.shape[1], device=x.device, dtype=torch.long)
         mask = self.causal_mask[None, None, input_pos]
         freqs_cis = self.freqs_cis[input_pos]
+
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply build/model_dist.py]: same as above - ran ruff format to make this uniform.

File: build/model_dist.py [hunk 21]
```diff
@@ -50,7 +52,7 @@ def __init__(self, config: TransformerArgs, stage_idx: int, n_stages: int) -> No
                 device_mesh,
                 RowwiseParallel(input_layouts=Replicate()),
             )
-
+
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply build/model_dist.py]: run ruff format on it to resolve

File: build/model_dist.py (hunk 1 above)
Comment (lessw2020) [reply build/model_dist.py]: thanks for catching.  I didn't intend to leave this in, and have removed.

File: build/model_dist.py (hunk 2 above)
Comment (lessw2020) [reply build/model_dist.py]: agree and removed.

File: dist_run.py (hunk 3 above)
Comment (lessw2020) [reply dist_run.py:19]: makes sense...added a TODO comment to mark this for future PR.

File: dist_run.py (hunk 7 above)
Comment (lessw2020) [reply dist_run.py:87]: good ideas but atm neither do what we need.  
1 - config does not contain model dtype
2 - tested the get precision api, and just returns float32 atm, so I've left a commented out line using it with a TODO to investigate.  It would be nice if this works so we don't have to maintain our own list. 

~~~
[rank0]:08-29 19:51:19.455 - __main__:65 - INFO - Chat Model Config: ModelArgs(text_transformer_args=TransformerArgs(block_size=2048, vocab_size=32000, n_layers=32, n_heads=32, dim=4096, hidden_dim=11008, n_local_heads=32, head_dim=128, rope_base=10000, norm_eps=1e-05, multiple_of=256, ffn_dim_multiplier=None, use_tiktoken=False, max_seq_length=8192, use_scaled_rope=False))
[rank0]:08-29 19:51:19.455 - __main__:67 - INFO - Using torchchat precision torch.float32
~~~

File: dist_run.py (hunk 10 above)
Comment (lessw2020) [reply dist_run.py:126]: I had that there intentionally earlier (to check pre and post) but you are right, we don't need it x2 now that things are working. 
Have removed to just leave the second one as a verification after loading.

File: distributed/safetensor_utils.py (hunk 18 above)
Comment (lessw2020) [reply distributed/safetensor_utils.py:300]: ah good catch - I try to doc-string every function.  Have added.

File: distributed/safetensor_utils.py (hunk 17 above)
Comment (lessw2020) [reply distributed/safetensor_utils.py:280]: good point, removed.

File: distributed/safetensor_utils.py (hunk 16 above)
Comment (lessw2020) [reply distributed/safetensor_utils.py:277]: ah, this is b/c I have two branches, one for dtensors and one for regular tensors...let me just consolidate it as we only need this last one and not the one special cased under regular tensors.

File: distributed/safetensor_utils.py (hunk 15 above)
Comment (lessw2020) [reply distributed/safetensor_utils.py:248]: we need this otherwise:
WARNING - Partially updated state dict. Missing 1 keys: {'output.weight'} 
if we don't have it.

File: distributed/dtensor_utils.py [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+                cpu_tensors.append(f"{mod.__class__.__name__}.weights")
+        
+
+        for name, child in mod.named_children():
+            recurse(child)
+
+    recurse(module)
+    return cpu_tensors
+
+def record_module_dtypes(module):
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply distributed/dtensor_utils.py:38]: I want to have it available to pop in dtype checks if needed, at least until we have proven out loading on a number of different models.  
Also, the get_precision/set_precision do not do what we thought atm.  
I only see them being used right now for setting the precision of the kvcache dtype, not the model dtypes.

File: distributed/dtensor_utils.py [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+from distributed.logging_utils import setup_logging
+from collections import defaultdict
+
+logger = setup_logging(__name__)
+
+def is_dtensor(tensor):
+    """ Check if a tensor is a DTensor by class or has a placements attribute (not sure if we want to use attr check) """
+    return isinstance(tensor, DTensor) or hasattr(tensor, 'placements')
+
+def find_cpu_tensors(module):
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply distributed/dtensor_utils.py:13]: good idea - let me do it in a future PR though as I want to get this landed for now. 
Added a TODO comment for it for now.

File: dist_run.py (hunk 8 above)
Comment (lessw2020) [reply dist_run.py:118]: yes -  have switched the order and verified all working.

File: distributed/dtensor_utils.py [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+
+    recurse(module)
+    return dtype_count, dtype_locations, fp32_locations
+
+
+def load_into_dtensor(weight_tensor, model_dtensor, debug=False):
+    """ Adjust a loaded tensor to match the shape/placement of the model DTensor and copy the data into it """
+    weight_tensor = weight_tensor.to(model_dtensor.device)
+    if debug:
+        logger.info(f"Loading into DTensor: {weight_tensor.shape=} {model_dtensor.shape=}")
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply distributed/dtensor_utils.py:73]: actually I just removed these - I needed it when writing this function but assuming no major dtensor changes probably won't need it again and the code is easier to read w/o it.

File: distributed/dtensor_utils.py [hunk 25]
```diff
@@ -0,0 +1,156 @@
...
+        raise ValueError(f"Shape mismatch: weight tensor shape {weight_tensor.shape} "
+                         f"doesn't match DTensor shape {model_dtensor.shape}")
+    
+    placements = model_dtensor.placements
+    mesh = model_dtensor.device_mesh
+    mesh_dims = mesh.ndim
+    
+    if debug:
+        logger.info(f"DTensor: Placements: {placements}")
+        logger.info(f"DTensor: Mesh: {mesh}, mesh_dims={mesh_dims}")
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply distributed/dtensor_utils.py:85]: removed

File: distributed/dtensor_utils.py [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+
+    for placement in placements:
+        if isinstance(placement, Shard):
+            shard_dim = placement.dim
+            if debug:
+                logger.info(f"DTensor Sharding dimension: {shard_dim}")
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply distributed/dtensor_utils.py:91]: removed

File: distributed/safetensor_utils.py (hunk 19 above)
Comment (lessw2020) [reply distributed/safetensor_utils.py:320]: moved it inline
